utils.py

```diff
@@ -1,19 +1,39 @@
-from collections import UserString
 import os
 from pathlib import Path
 import sys
 import ast
 from dataclasses import dataclass
+from types import NotImplementedType
 from typing import Callable, Dict, List
 from filecmp import cmp
-from typing_extensions import Self
 
 ################################################################################
 # Labels
 ################################################################################
-class Label(str):
-    def __new__(cls: type[Self], seq: object) -> Self:
-        return str.__new__(cls, "_" + seq if sys.platform == "darwin" else seq)
+LABEL_PREPEND: str = "_" if sys.platform == "darwin" else ""
+
+
+@dataclass(frozen=True)
+class Label:
+    name: str
+
+    def __init__(self, _name: str) -> None:
+        object.__setattr__(self, "name", LABEL_PREPEND + _name)
+
+    def __add__(self, other) -> str | NotImplementedType:
+        if isinstance(other, str):
+            return self.name + other
+        else:
+            return NotImplemented
+
+    def __radd__(self, other) -> str | NotImplementedType:
```

Comment:
You don't want an __radd__ method because it destroys the prefix. 
It shouldn't be used for labels, anyway. (is it?)

Reply:
There are a bunch of string & Label concatenations throughout the library. Both `__add__` and `__radd__` are only implemented for concatenations of `Label`s with strings, so the return type will only be a string. I thought this would be okay, as one can't implicitly create a messed up `Label`.

If you think this should definitely not stay as is, I can:

a) implement `Label` + `Label` case with platform prefix in mind, or
b) remove it altogether and do explicit `str()` castings wherever required.

Which one would you prefer?

Reply:
I see. I didn't realize the return type was a string.
That sounds ok to me. Let's see how this works for the students.